src/attributes.rs

```diff
@@ -0,0 +1,282 @@
+use std::rc::Rc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Attribute<'a> {
+    pub is_inner: bool,
+    pub content: Rc<AttributeMetaItem<'a>>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_item
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_trimmed = raw.trim();
+        let raw_without_closing = raw_trimmed.strip_suffix(']').unwrap_or_else(|| {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it is not closed with a square bracket."
+                ),
+                raw_trimmed
+            )
+        });
+
+        if let Some(raw_content) = raw_without_closing.strip_prefix("#[") {
+            Attribute {
+                is_inner: false,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else if let Some(raw_content) = raw_without_closing.strip_prefix("#![") {
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it starts with neither `#[` nor `#![`."
+                ),
+                raw_trimmed
+            )
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AttributeMetaItem<'a> {
+    pub raw_item: &'a str,
+    pub base: &'a str,
+    pub assigned_item: Option<&'a str>,
+    pub arguments: Option<Vec<Rc<AttributeMetaItem<'a>>>>,
+}
+
+impl<'a> AttributeMetaItem<'a> {
+    /// Tries to parse `raw` as a comma-separated sequence of `AttributeMetaItem`'s
+    /// wrapped in parentheses, square brackets or curly brackets.
+    fn slice_arguments(raw: &'a str) -> Option<Vec<Rc<AttributeMetaItem<'a>>>> {
+        let is_left = |c| c == '(' || c == '[' || c == '{';
+        let is_right = |c| c == ')' || c == ']' || c == '}';
+        let matching_right = |c| match c {
+            '(' => ')',
+            '[' => ']',
+            '{' => '}',
+            _ => unreachable!("Tried to find matching right bracket for {}.", c),
```

Comment:
```suggestion
            _ => unreachable!("Tried to find matching right bracket for {c}."),
```
Also because of that clippy lint I mentioned earlier. Try to replace all the other similar uses as well.